File: ios/Earwyrm/Core/Models/AppNotification.swift
import Foundation

struct AppNotification: Codable, Identifiable {
    let id: UUID
    let userId: UUID
    let type: String
    let lyricId: UUID?
    let commentId: UUID?
    let collectionId: UUID?
    let actorId: UUID?
    let lyricSnippet: String?
    let songTitle: String?
    let artistName: String?
    let shareToken: String?
    let actorUsername: String?
    let commentContent: String?
    let collectionName: String?
    let followType: String?
    let followValue: String?
    let readAt: Date?
    let createdAt: Date

    enum CodingKeys: String, CodingKey {
        case id
        case userId = "user_id"
        case type
        case lyricId = "lyric_id"
        case commentId = "comment_id"
        case collectionId = "collection_id"
        case actorId = "actor_id"
        case lyricSnippet = "lyric_snippet"
        case songTitle = "song_title"
        case artistName = "artist_name"
        case shareToken = "share_token"
        case actorUsername = "actor_username"
        case commentContent = "comment_content"
        case collectionName = "collection_name"
        case followType = "follow_type"
        case followValue = "follow_value"
        case readAt = "read_at"
        case createdAt = "created_at"
    }
}

/// Encodable struct for app-side share notification INSERT.
struct ShareNotificationInsert: Encodable {
    let userId: UUID
    let type: String
    let lyricId: UUID
    let actorId: UUID
    let lyricSnippet: String?
    let songTitle: String?
    let artistName: String?
    let shareToken: String?
    let actorUsername: String?

    enum CodingKeys: String, CodingKey {
        case userId = "user_id"
        case type
        case lyricId = "lyric_id"
        case actorId = "actor_id"
        case lyricSnippet = "lyric_snippet"
        case songTitle = "song_title"
        case artistName = "artist_name"
        case shareToken = "share_token"
        case actorUsername = "actor_username"
    }
}

File: android/app/src/main/java/com/earwyrm/app/core/model/AppNotification.kt
package com.earwyrm.app.core.model

import kotlinx.datetime.Instant
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class AppNotification(
    val id: String, @SerialName("user_id") val userId: String, val type: String,
    @SerialName("lyric_id") val lyricId: String? = null,
    @SerialName("comment_id") val commentId: String? = null,
    @SerialName("collection_id") val collectionId: String? = null,
    @SerialName("actor_id") val actorId: String? = null,
    @SerialName("lyric_snippet") val lyricSnippet: String? = null,
    @SerialName("song_title") val songTitle: String? = null,
    @SerialName("artist_name") val artistName: String? = null,
    @SerialName("share_token") val shareToken: String? = null,
    @SerialName("actor_username") val actorUsername: String? = null,
    @SerialName("comment_content") val commentContent: String? = null,
    @SerialName("collection_name") val collectionName: String? = null,
    @SerialName("follow_type") val followType: String? = null,
    @SerialName("follow_value") val followValue: String? = null,
    @SerialName("read_at") val readAt: Instant? = null,
    @SerialName("created_at") val createdAt: Instant
)

@Serializable
data class ShareNotificationInsert(
    @SerialName("user_id") val userId: String,
    val type: String = "share",
    @SerialName("lyric_id") val lyricId: String,
    @SerialName("actor_id") val actorId: String,
    @SerialName("lyric_snippet") val lyricSnippet: String? = null,
    @SerialName("song_title") val songTitle: String? = null,
    @SerialName("artist_name") val artistName: String? = null,
    @SerialName("share_token") val shareToken: String? = null,
    @SerialName("actor_username") val actorUsername: String? = null
)
